File: 2816-double-a-number-represented-as-a-linked-list/2816-double-a-number-represented-as-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode doubleIt(ListNode head) {
        head = reverse(head);
        ListNode digit = head;
        boolean carry = false;
        while(digit != null){
            int num = digit.val;
            num *= 2;
            if(carry){
                num++;
                carry = false;
            } 
            digit.val = num%10;
            if(num > 9){
                carry = true;
            }
            if(digit.next == null) break;
            digit = digit.next;
            
        }
        if(carry){
            ListNode cry = new ListNode(1);
            digit.next = cry;
        }
        return reverse(head);
         
    }

    public ListNode reverse(ListNode head){
        ListNode curr = head;
        ListNode prev = null;
        ListNode next;
        while(curr != null){
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
}

File: 2816-double-a-number-represented-as-a-linked-list/2816-double-a-number-represented-as-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* doubleIt(ListNode* head) {
        head = reverse(head);
        ListNode* digit = head;
        bool carry = false;
        while(digit != NULL){
            int num = digit->val;
            num *= 2;
            if(carry){
                num++;
                carry = false;
            } 
            digit->val = num%10;
            if(num > 9){
                carry = true;
            }
            if(digit->next == NULL) break;
            digit = digit->next;
            
        }
        if(carry){
            ListNode* cry = new ListNode(1);
            digit->next = cry;
        }
        return reverse(head);
         
    }

    ListNode* reverse(ListNode* head){
        ListNode* curr = head;
        ListNode* prev = NULL;
        ListNode* next;
        while(curr != NULL){
            next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
};